x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java

```diff
@@ -198,37 +190,71 @@ protected void taskOperation(StopDataFrameTransformAction.Request request, DataF
                     // Wait until the persistent task is stopped
                     // Switch over to Generic threadpool so we don't block the network thread
                     threadPool.generic().execute(() ->
-                        waitForDataFrameStopped(request.getExpandedIds(), request.getTimeout(), onStopListener));
+                        waitForDataFrameStopped(request.getExpandedIds(), request.getTimeout(), request.isForce(), onStopListener));
                 },
                 listener::onFailure
         );
     }
 
-    private void waitForDataFrameStopped(Collection<String> persistentTaskIds, TimeValue timeout,
-                                         ActionListener<StopDataFrameTransformAction.Response> listener) {
+    private void waitForDataFrameStopped(Set<String> persistentTaskIds,
+                                         TimeValue timeout,
+                                         boolean force,
+                                         ActionListener<Response> listener) {
+        // This map is accessed in the predicate and the listener callbacks
+        final Map<String, ElasticsearchException> exceptions = new ConcurrentHashMap<>();
         persistentTasksService.waitForPersistentTasksCondition(persistentTasksCustomMetaData -> {
-
             if (persistentTasksCustomMetaData == null) {
                 return true;
             }
-
             for (String persistentTaskId : persistentTaskIds) {
-                if (persistentTasksCustomMetaData.getTask(persistentTaskId) != null) {
-                    return false;
+                PersistentTasksCustomMetaData.PersistentTask<?> transformsTask = persistentTasksCustomMetaData.getTask(persistentTaskId);
+                // Either the task has successfully stopped or we have seen that it has failed
+                if (transformsTask == null || exceptions.containsKey(persistentTaskId)) {
+                    continue;
+                }
+
+                // If force is true, then it should eventually go away, don't add it to the collection of failures.
+                DataFrameTransformState taskState = (DataFrameTransformState)transformsTask.getState();
+                if (force == false && taskState != null && taskState.getTaskState() == DataFrameTransformTaskState.FAILED) {
+                    exceptions.put(persistentTaskId, new ElasticsearchStatusException(
+                        DataFrameMessages.getMessage(DATA_FRAME_CANNOT_STOP_FAILED_TRANSFORM,
+                            persistentTaskId,
+                            taskState.getReason()),
+                        RestStatus.CONFLICT));
+
+                    // If all the remaining tasks are flagged as failed, do not wait for another ClusterState update.
+                    // Return to the caller as soon as possible
+                    return persistentTasksCustomMetaData.tasks().stream().allMatch(p -> exceptions.containsKey(p.getId()));
```

Comment:
I noticed this occur in my local tests. If the clusterstate is not updated for 30s, this predicate times out without this check because it never sees that all the tasks have been flagged as failed.